### PR TITLE
feat(llma): replace sklearn HDBSCAN with hdbscan-rs 10x faster clustering

### PR DESCRIPTION
**Pull request by JasonLovesDoggo**
## Summary
Hey! I've been an early beta user of the LLM trace clustering feature, and I
just saw [the blog post](https://posthog.com/blog/llm-analytics-clustering-how-it-works) about how the clustering pipeline works. I've been working on a Rust-based HDBSCAN
implementation ([hdbscan-rs](https://github.com/JasonLovesDoggo/hdbscan-rs)) over the past couple of days, so I thought I'd share the performance gains.

It's a one-line import swap - hdbscan-rs has been built from the start to be
100% compatible with scikit-learn's API, so nothing else needs to change.

## Benchmarks

| Samples | Features | hdbscan-rs | sklearn | Speedup |
|---------|----------|-----------|---------|---------|
| 100 | 10 | 0.0004s | 0.0017s | **4.6x** |
| 500 | 50 | 0.0031s | 0.0175s | **5.6x** |
| 1,000 | 100 | 0.0133s | 0.1389s | **10.4x** | 
| 2,000 | 100 | 0.0516s | 0.5140s | **10.0x** |
| 5,000 | 50 | 0.2656s | 1.1006s | **4.1x** |

Worth highlighting the 100-feature rows specifically — that matches your
`DEFAULT_UMAP_N_COMPONENTS = 100` in prod. At 1k samples/100d you're looking
at ~10x faster, and at 2k samples/100d it's still ~10x. 

## Changes
- `clustering.py`: `from sklearn.cluster import HDBSCAN` → `from hdbscan_rs import HDBSCAN`
- `pyproject.toml`: add `hdbscan-rs>=0.5.0` dependency

## Test plan
- [x] All 17 existing `test_clustering.py` tests pass unchanged
- [x] Cluster count parity verified against sklearn on synthetic data
- [x] Probabilities, centroids, and label contiguity all validated
- [ ] CI for downstream workflow/integration tests

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -116,6 +116,7 @@ dependencies = [
     "requests~=2.32.3",
     "retry==0.9.2",
     "s3fs==2025.9.0",
+    "hdbscan-rs>=0.5.0",
```
[the remark this answers is not in the document]
Comment (JasonLovesDoggo) [reply pyproject.toml]: This is completely fair. I have pinned hdbscan-rs to 0.5.0 here but if you have any other suggestions, I am happy to change this PR to help mitigate the supply chain risk issue.

File: posthog/temporal/llm_analytics/trace_clustering/clustering.py [hunk 2]
```diff
@@ -20,7 +20,8 @@
 from typing import TYPE_CHECKING, Literal
 
 import numpy as np
-from sklearn.cluster import HDBSCAN, KMeans
+from hdbscan_rs import HDBSCAN
+from sklearn.cluster import KMeans
```
[the remark this answers is not in the document]
Comment (JasonLovesDoggo) [reply posthog/temporal/llm_analytics/trace_clustering/clustering.py:24]: hdbscan-rs supports both `cluster_selection_method="eom"` (raises ValueError for invalid values, doesn't silently ignore) and `probabilities_` after fit_predict(). Happy to add a parity test that runs both implementations on the same input and asserts matching outputs if desired.

**Comment by Twixes** [comment]
Hey @JasonLovesDoggo! I'm not on the LLM analytics team, but I've worked with clustering of embeddings too, having used https://pypi.org/project/fast-hdbscan/ as the faster alternative to scikit.
Do you have a benchmark against `fast-hdbscan`?

File: pyproject.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewm4894) [reply pyproject.toml]: haha i was going to mention this side of things! defo going to dig in here though and have a good look

**Comment by andrewm4894** [comment]
sorry im so slow getting to this one. i do think its very unlikely we would just adopt a new dep like this unless/untill we actually have users triggering on demand clustering and then kinda waiting in the loop where latency matters more.

@PostHog/team-replay - is clustering speed a bottleneck for you folks at all for any clustering in the session summary stuff?

**Comment by JasonLovesDoggo** [comment]
> sorry im so slow getting to this one. i do think its very unlikely we would just adopt a new dep like this unless/untill we actually have users triggering on demand clustering and then kinda waiting in the loop where latency matters more.
> 
> @PostHog/team-replay - is clustering speed a bottleneck for you folks at all for any clustering in the session summary stuff?

All good I get it. I look forward to the day where you guys are bottlenecked by clustering haha

**Comment by Twixes** [comment]
I think we'll close this specific PR for now, but we'll keep hdbscan-rs in mind for future clustering work! Thank you @JasonLovesDoggo

**Comment by JasonLovesDoggo** [comment]
> I think we'll close this specific PR for now, but we'll keep hdbscan-rs in mind for future clustering work! Thank you @JasonLovesDoggo

All good! I completely understand. Until the next PR! 👋🏼
